src/blas/impl/KokkosBlas_Newton_impl.hpp

```diff
@@ -0,0 +1,240 @@
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef __KOKKOSBATCHED_ODE_NEWTON_HPP__
+#define __KOKKOSBATCHED_ODE_NEWTON_HPP__
+
+#include "Kokkos_Core.hpp"
+#include "KokkosBatched_LU_Decl.hpp"
+#include "KokkosBatched_LU_Serial_Impl.hpp"
+#include "KokkosBatched_Gesv.hpp"
+#include "KokkosBlas1_nrm2.hpp"
+#include "KokkosBlas1_scal.hpp"
+#include "KokkosBlas1_axpby.hpp"
+
+namespace KokkosBlas {
+namespace Impl {
+
+enum class NewtonSolverStatus { Converged = 0, LinearSolveFailure, MaxIters };
+
+std::ostream& operator<<(std::ostream& os, NewtonSolverStatus& status) {
+  switch (status) {
+    case NewtonSolverStatus::Converged: os << "Newton Solver Converged!"; break;
+    case NewtonSolverStatus::LinearSolveFailure:
+      os << "Newton: Linear Solver Failure";
+      break;
+    case NewtonSolverStatus::MaxIters:
+      os << "Newton reached maximum iterations without convergence.";
+      break;
+  }
+  return os;
+}
+
+/// \brief NewtonHandle
+///
+/// This handle is used to pass information between the Newton Solver and
+/// the calling code.
+///
+/// \tparam: NormViewType: Type of view used to store the residual convergence
+/// history
+
+template <class NormViewType>
+struct NewtonHandle {
+  using norm_type = typename NormViewType::non_const_value_type;
+
+  NormViewType lastResidual;  // Residual of last successful iteration
+  typename NormViewType::HostMirror lastResidualHost;
+
+  // NormViewType  residual_norms;
+  // TODO: Making these public for now. Should make private and access
+  // via setters and getters?
+  int maxIters;           // Maximum number of Newton steps
+  norm_type relativeTol;  // Relative convergence tolerance
+  bool debug_mode;        // Returns extra verbose output if true.
+
+  NewtonHandle(int _maxIters = 25, double _relativeTol = 1.0e-6,
+               bool _debug = false)
+      : lastResidual("ending Residual norm", 1),
+        lastResidualHost("end res norm host", 1),
+        maxIters(_maxIters),
+        relativeTol(_relativeTol),
+        debug_mode(_debug) {}
+
+  KOKKOS_FUNCTION
+  void set_residual(const norm_type val) const { lastResidual(0) = val; }
+
+  KOKKOS_FUNCTION
+  norm_type get_residual() const { return lastResidual(0); }
+
+  norm_type get_residual_host() const {
+    Kokkos::deep_copy(lastResidualHost, lastResidual);
+    return lastResidualHost(0);
+  }
+
+};  // NewtonHandle
+
+/// \brief Newton Functor:
+/// Solves the nonlinear system F(x) = 0
+/// where F is a map from R^n to R^n.
+/// \tparam System: Struct that allows the evaluation
+///         of the residual and jacobian using the
+///         residual() and jacobian() methods.
+/// \tparam Matrix: rank-2 view-type
+/// \tparam XVector: rank-1 view-type
+/// \tparam YVector: rank-1 view-type
+/// \param
+/// \param X [in]: Input vector X, a rank 1 view
+/// \param Y [in/out]: Output vector Y, a rank 1 view
+///
```

Comment:
We are solving `F(x)=0` by linearizing it into `J*x=-y`, `y` is a work vector that stores the right hand side, it's actually called `rhs` in the functor implementation. In short we do this:
```
sys.residual(x, y);  // compute residual and store in y
sys.jacobian(x, J) // compute the jacobian matrix J
Gesv(J, update, rhs, tmp);
update = -update; // because we really want to solver J*update = -rhs
x=x+update;
```